### PR TITLE
Core/Creature: implement school immunity for creatures in creature_template

### PR DESCRIPTION
**Pull request by Wyreth**
**Changes proposed:**

Following the discussion in #20464, creatures need a way to be immune to certain spell schools without the use of specific auras.

This copies the existing handling for mechanic immunity and applies it to spell immunity.

**Target branch(es):**

- [x] 3.3.5
- [x] master

**Tests performed:** works fine.

### CODE REVIEW

**Comment by ccrs** [comment]
is there any sort of list of creatures using this?

> without the use of specific auras.
that feels like a short list to me

**Comment by Wyreth** [comment]
offl posted one [https://github.com/TrinityCore/TrinityCore/pull/20464#issuecomment-333293467](here). Checked most of them (haven't finished yet), seems complete.

It's not a lot, but adding auras to them seems hacky and there's probably some missing (all/almost all Vanilla/TBC elementals are immune to their own elements).

**Comment by ccrs** [comment]
> but adding auras to them seems hacky and there's probably some missing

sniffs are the way to go then, if possible

> all/almost all Vanilla/TBC elementals are immune to their own elements, for example

many of these removed on wotlk transition

the link points to nothing btw

**Comment by Wyreth** [comment]
Fixed the link, sorry :P

I recall most immunities were removed in Cataclysm with the old world rehaul, and only some were affected in WotLK (notable one: Onyxia, but only because of the 3.2 rehash). This needs a bit more research, but it's part of the "scripting" process.

All the creatures involved are part of the old world that was removed since Cataclysm, getting sniffs for most of them is probably impossible by now. Still we have a list, so if anyone is willing to check existing archives for auras, they're welcome to :P

**Comment by offl** [comment]
You can check sniffs for high(TBC) entries if there is any aura. Entries: 16488,16491,16530,16854,18099,18100,18101,18102,18864,18865,18866,18867,19205,20516,20702,21032,22310.

File: src/server/game/Globals/ObjectMgr.cpp [hunk 1]
```diff
@@ -414,8 +414,8 @@ void ObjectMgr::LoadCreatureTemplates()
                                              "spell2, spell3, spell4, spell5, spell6, spell7, spell8, PetSpellDataId, VehicleId, mingold, maxgold, AIName, MovementType, "
     //                                        62           63           64              65            66             67              68                  69
                                              "InhabitType, HoverHeight, HealthModifier, ManaModifier, ArmorModifier, DamageModifier, ExperienceModifier, RacialLeader, "
-    //                                             70          71                72               73          74
-                                             "movementId, RegenHealth, mechanic_immune_mask, flags_extra, ScriptName "
+    //                                             70          71                72                    73                  74           75
```
Comment (tkrokli) [inline src/server/game/Globals/ObjectMgr.cpp]: I know this is cosmetic (comment), but if you look at the numbering before 70, you will see that the numbers are placed at the same column as the first letter in the field name. Would you mind aligning the numbers to the previous (untouched) lines, even if this is just a comment line?

Comment (Wyreth) [reply src/server/game/Globals/ObjectMgr.cpp]: Sure! Didn't notice it at first. Thanks.

**Comment by k59** [comment]
Fwiw, "World of Warcraft Dungeon Companion" book mentions immunities.

School immunities(templates?):
https://i.imgur.com/QTvTfOH.png
https://i.imgur.com/r93Y0MT.png
https://i.imgur.com/EMVXKDZ.png
https://i.imgur.com/3FfGEi6.png (aura?)

Flag immunities:
https://i.imgur.com/UYfBybs.png
https://i.imgur.com/derhICi.png


Scanned PDF: https://www.scribd.com/document/343312627/World-of-Warcraft-Dungeon-Companion

**Comment by offl** [comment]
This book looks great :)

**Comment by offl** [comment]
To check(probably some NPCs should be added to the list i provided)
https://github.com/cmangos/wotlk-db/commit/d7a85682b572480bdcbd6586d8057f7380e4dd83
https://github.com/cmangos/wotlk-db/commit/2a01a9ef7a42ac585cd5e5c4d8bb75859912ec70
https://github.com/cmangos/wotlk-db/commit/6f86aaa63af0c40544e5f6f739275bdca52f80cb
https://github.com/cmangos/wotlk-db/commit/ab534ed05d9c9ce1ca3c508f75dafb8e5a936643#diff-8e1f984f063555034f8244c7872cd836R265
https://github.com/cmangos/wotlk-db/commit/b2dcc777eac0aa1960448880f502aeecd7ee9c6a#diff-b2a8306a5add8b5a5539cf96fce89564R5
https://github.com/cmangos/wotlk-db/commit/911c8b47c0699bd28eb768222fe1b34cdf91caec#diff-e1b68a3dce525062a38ca4aec08b9674R3
https://github.com/cmangos/wotlk-db/commit/5cb903d5806a81ad1047d7dbfe5807edf5143d97#diff-5340131c5c631e90fce5f352fd081bc3R315

**Comment by offl** [comment]
Final version with CMANGOS code:
```sql
(12457,16491,18099,18100,18101,18102); -- Holy
(2091,2108,2447,2726,2760,4036,4037,4038,4676,5850,5852,7137,7664,7734,7735,9016,9017,9026,9816,9878,9879,10485,11502,11666,11667,11668,12056,12143,12457,16491,18099,18100,18101,18102,19203,19973,20703,21706,22323,25001); -- Fire
(92,2258,2592,2735,2736,2752,2762,2791,4499,10485,11321,11744,11777,11778,11779,11783,11784,12201,12237,12457,14435,14455,16491,17085,17154,17156,17157,17158,17159,17160,18062,18099,18100,18101,18102,18181,21707,22036,23029); -- Nature
(691,2761,3917,5461,5462,7079,7132,7428,7429,7664,7734,7735,9453,10198,10485,10642,10756,10757,12457,13280,16491,17153,17155,17167,17207,17358,18099,18100,18101,18102,19204,20704,21428,21728,22009,22035,22309,25740,25755,25756,25757,25865,26116,26178,26204,26214,26215,26216,26338,26339,26340,26341,26342); -- Frost
(703,1043,1364,7664,10485,12457,16491,18099,18100,18101,18102,19206,20705,21032); -- Shadow
(7734,7735,10202,10485,10662,10663,10664,11480,11483,12457,16488,16491,16530,16854,18099,18100,18101,18102,18864,18865,18866,18867,19205,20516,20702,21032,22310); -- Arcane

UPDATE `creature_template` SET `dmgschool` = 4 WHERE `entry` = 17153; -- Lake Spirit
UPDATE `creature_template` SET `dmgschool` = 3 WHERE `entry` = 17154; -- Muck Spawn
UPDATE `creature_template` SET `dmgschool` = 4 WHERE `entry` = 17155; -- Lake Surger
```

CMANGOS source used:
https://github.com/cmangos/wotlk-db/commit/d7a85682b572480bdcbd6586d8057f7380e4dd83
https://github.com/cmangos/wotlk-db/commit/b2dcc777eac0aa1960448880f502aeecd7ee9c6a#diff-b2a8306a5add8b5a5539cf96fce89564R5
https://github.com/cmangos/wotlk-db/commit/5cb903d5806a81ad1047d7dbfe5807edf5143d97#diff-5340131c5c631e90fce5f352fd081bc3R315

NPCs 10641,7136,5354,5833 have immunities on CMANGOS, but I can't find any source for them.

**Comment by offl** [comment]
And some WotLK elementals:
```sql
(16570,24228,24601,25226,25514,25715,26316,30846); -- Frost
(25417,30847); -- Fire
(26370,30848); -- Arcane
(26407,28411,28585,28825); -- Nature
```

File: src/server/game/Entities/Creature/Creature.cpp [hunk 2]
```diff
@@ -2052,6 +2053,26 @@ void Creature::LoadMechanicTemplateImmunity()
     }
 }
 
+void Creature::LoadSpellTemplateImmunity()
+{
+    // uint32 max used for "spell id", the immunity system will not perform SpellInfo checks against invalid spells
+    // used so we know which immunities were loaded from template
+    static uint32 const placeholderSpellId = std::numeric_limits<uint32>::max();
```
Comment (jackpoz) [inline src/server/game/Entities/Creature/Creature.cpp]: I would just keep it simple and skip the "static const" part

**Comment by jackpoz** [comment]
@Wyreth is this PR ready to be merged ? it looks ready to me (I personally would remove the "static const" for the spell id placeholder and just declare it as a simple variable, just personal taste).

**Comment by offl** [comment]
I think last time he was online one month ago.

**Comment by tkrokli** [comment]
Let us hope he will be back during the holidays.
Otherwise, we may have to update it ourselves.

**Comment by jackpoz** [comment]
whoever knows about which immunities to include, please follow the usual procedure for SQL script fixes

**Comment by offl** [comment]
Dunno how SQL should look like.
```sql
UPDATE `creature_template` SET `spell_school_immune_mask`=`spell_school_immune_mask`| ?
```

**Comment by jackpoz** [comment]
Use the values defined at https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Miscellaneous/SharedDefines.h#L252

```
enum SpellSchoolMask
{
    SPELL_SCHOOL_MASK_NONE    = 0x00,                       // not exist
    SPELL_SCHOOL_MASK_NORMAL  = (1 << SPELL_SCHOOL_NORMAL), // PHYSICAL (Armor)
    SPELL_SCHOOL_MASK_HOLY    = (1 << SPELL_SCHOOL_HOLY),
    SPELL_SCHOOL_MASK_FIRE    = (1 << SPELL_SCHOOL_FIRE),
    SPELL_SCHOOL_MASK_NATURE  = (1 << SPELL_SCHOOL_NATURE),
    SPELL_SCHOOL_MASK_FROST   = (1 << SPELL_SCHOOL_FROST),
    SPELL_SCHOOL_MASK_SHADOW  = (1 << SPELL_SCHOOL_SHADOW),
    SPELL_SCHOOL_MASK_ARCANE  = (1 << SPELL_SCHOOL_ARCANE),
```

i.e. SPELL_SCHOOL_MASK_ARCANE is 64 or 0x40

**Comment by offl** [comment]
```sql
UPDATE `creature_template` SET `spell_school_immune_mask`=`spell_school_immune_mask`|2 WHERE `entry` IN (12457,16491,18099,18100,18101,18102); -- Holy (actually 18099,18100,18101,18102 should be immune to any mechanic and damage(melee and magic))
UPDATE `creature_template` SET `spell_school_immune_mask`=`spell_school_immune_mask`|4 WHERE `entry` IN (2091,2108,2447,2726,2760,4036,4037,4038,4676,5850,5852,7137,7664,7734,7735,9016,9017,9026,9816,9878,9879,10485,11502,11666,11667,11668,12056,12143,12457,16491,18099,18100,18101,18102,19203,19973,20703,21706,22323,25001,25417,30847); -- Fire
UPDATE `creature_template` SET `spell_school_immune_mask`=`spell_school_immune_mask`|8 WHERE `entry` IN (92,2258,2592,2735,2736,2752,2762,2791,4499,10485,11321,11744,11777,11778,11779,11783,11784,12201,12237,12457,14435,14455,16491,17085,17154,17156,17157,17158,17159,17160,18062,18099,18100,18101,18102,18181,21707,22036,23029,26407,28411,28585,28825); -- Nature
UPDATE `creature_template` SET `spell_school_immune_mask`=`spell_school_immune_mask`|16 WHERE `entry` IN (691,2761,3917,5461,5462,7079,7132,7428,7429,7664,7734,7735,9453,10198,10485,10642,10756,10757,12457,13280,16491,17153,17155,17167,17207,17358,18099,18100,18101,18102,19204,20704,21428,21728,22009,22035,22309,25740,25755,25756,25757,25865,26116,26178,26204,26214,26215,26216,26338,26339,26340,26341,26342,16570,24228,24601,25226,25514,25715,26316,30846); -- Frost
UPDATE `creature_template` SET `spell_school_immune_mask`=`spell_school_immune_mask`|32 WHERE `entry` IN (703,1043,1364,7664,10485,12457,16491,18099,18100,18101,18102,19206,20705,21032); -- Shadow
UPDATE `creature_template` SET `spell_school_immune_mask`=`spell_school_immune_mask`|64 WHERE `entry` IN (7734,7735,10202,10485,10662,10663,10664,11480,11483,12457,16488,16491,16530,16854,18099,18100,18101,18102,18864,18865,18866,18867,19205,20516,20702,21032,22310,26370,30848); -- Arcane

UPDATE `creature_template` SET `dmgschool` = 4 WHERE `entry` = 17153; -- Lake Spirit
UPDATE `creature_template` SET `dmgschool` = 3 WHERE `entry` = 17154; -- Muck Spawn
UPDATE `creature_template` SET `dmgschool` = 4 WHERE `entry` = 17155; -- Lake Surger

**Comment by Killyana** [comment]
Better to test on retail: (actually 18099,18100,18101,18102 should be immune to any mechanic and damage (melee and magic)) before pushing this.

**Comment by jackpoz** [comment]
Maybe post the sql in another issue, this PR has been merged already and I'm not gonna look here for sql scripts to be applied

**Comment by offl** [comment]
> Better to test on retail: (actually 18099,18100,18101,18102 should be immune to any mechanic and damage (melee and magic)) before pushing this.

After reading the comments I'm not sure anymore that they should be immune to melee and magic damage. Removed.

**Comment by Killyana** [comment]
http://www.wowhead.com/npc=18102/kalandrios#comments:id=984751
Probably UNIT_FLAG_NON_ATTACKABLE

**Comment by offl** [comment]
They were sniffed, so there is a huge chance that their flags are correct. If someone know that current flags or something else is not correct open up an issue.

**Pull request merged**